girder/molecules/molecules/models/calculation.py

```diff
@@ -81,6 +87,76 @@ def validate(self, doc):
 
         return doc
 
+    def findcal(self, moleculeId=None, imageName=None,
+            inputParametersHash=None, inputGeometryHash=None,
+            name=None, inchi=None, inchikey=None, smiles=None, formula=None,
+            creatorId=None, pending=None, limit=None, offset=None,
+            sort=None, user=None):
+        # Set these to their defaults if they are not already set
+        limit, offset, sort = default_pagination_params(limit, offset, sort)
+
+        query = {}
+
+        # If a molecule id is specified it has higher priority
+        if moleculeId:
+            query['moleculeId'] = ObjectId(moleculeId)
+        # Otherwise, if query parameters for the molecules are specified, search for matching molecules first
+        elif any((name, inchi, inchikey, smiles, formula)):
+            params = {'offset': 0, 'limit': sys.maxsize}
+
+            if name:
+                params['name'] = name
+            if inchi:
+                params['inchi'] = inchi
+            if inchikey:
+                params['inchikey'] = inchikey
+            if smiles:
+                params['smiles'] = smiles
+            if formula:
+                params['formula'] = formula
+
+            molecules = MoleculeModel().findmol(params)['results']
+            moleculeIds = [molecule['_id'] for molecule in molecules]
+            query['moleculeId'] = {'$in': moleculeIds}
```

Comment:
We can probably speed this up if necessary with an aggregation pipeline, when we get a large dataset imported we can see how it performs.